File: lib/python/opentoken/src/test/opentoken/attributes/validation/not_null_or_empty_validator_test.py
"""
Tests for NotNullOrEmptyValidator.
"""

import pickle
import pytest

from opentoken.attributes.validation.not_null_or_empty_validator import NotNullOrEmptyValidator


class TestNotNullOrEmptyValidator:
    """Test cases for NotNullOrEmptyValidator."""

    def test_valid_values(self):
        """Test valid values."""
        validator = NotNullOrEmptyValidator()

        assert validator.eval("test") is True
        assert validator.eval("123") is True
        assert validator.eval("  test  ") is True
        assert validator.eval("special!@#$") is True
        assert validator.eval("multi\nline") is True

    def test_invalid_values(self):
        """Test invalid values."""
        validator = NotNullOrEmptyValidator()

        assert validator.eval(None) is False, "Null value should not be allowed"
        assert validator.eval("") is False, "Empty value should not be allowed"
        assert validator.eval(" ") is False, "Blank value should not be allowed"
        assert validator.eval("\t") is False, "Tab value should not be allowed"
        assert validator.eval("\n") is False, "Newline value should not be allowed"
        assert validator.eval("\r") is False, "Carriage return value should not be allowed"

    def test_serialization_should_preserve_state(self):
        """Test serialization and deserialization of the validator."""
        original_validator = NotNullOrEmptyValidator()

        # Serialize the validator using pickle
        serialized_data = pickle.dumps(original_validator)

        # Deserialize the validator
        deserialized_validator = pickle.loads(serialized_data)

        # Test that both validators behave the same way
        test_values = ["valid", None, "", " ", "\t", "\n", "test123"]

        for value in test_values:
            original_result = original_validator.eval(value)
            deserialized_result = deserialized_validator.eval(value)

            assert original_result == deserialized_result, (
                f"Validators should behave identically for value: {value} "
                f"(original: {original_result}, deserialized: {deserialized_result})"
            )

File: lib/java/opentoken/src/test/java/com/truveta/opentoken/attributes/validation/NotNullOrEmptyValidatorTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.validation;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;

import org.junit.jupiter.api.Test;

class NotNullOrEmptyValidatorTest {

    @Test
    void validTests() {
        var validator = new NotNullOrEmptyValidator();

        var result = validator.eval("test");
        assertTrue(result);

        result = validator.eval("123");
        assertTrue(result);

        result = validator.eval("  test  ");
        assertTrue(result);

        result = validator.eval("special!@#$");
        assertTrue(result);

        result = validator.eval("multi\nline");
        assertTrue(result);
    }

    @Test
    void invalidTests() {
        var validator = new NotNullOrEmptyValidator();

        var result = validator.eval(null);
        assertFalse(result, "Null value should not be allowed");
        result = validator.eval("");
        assertFalse(result, "Empty value should not be allowed");
        result = validator.eval(" ");
        assertFalse(result, "Blank value should not be allowed");
        result = validator.eval("\t");
        assertFalse(result, "Tab value should not be allowed");
        result = validator.eval("\n");
        assertFalse(result, "Newline value should not be allowed");
        result = validator.eval("\r");
        assertFalse(result, "Carriage return value should not be allowed");
    }

    @Test
    void serialization_ShouldPreserveState() throws Exception {
        // Test serialization and deserialization
        NotNullOrEmptyValidator originalValidator = new NotNullOrEmptyValidator();

        // Serialize the validator
        ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
        ObjectOutputStream out = new ObjectOutputStream(byteOut);
        out.writeObject(originalValidator);
        out.close();

        // Deserialize the validator
        ByteArrayInputStream byteIn = new ByteArrayInputStream(byteOut.toByteArray());
        ObjectInputStream in = new ObjectInputStream(byteIn);
        NotNullOrEmptyValidator deserializedValidator = (NotNullOrEmptyValidator) in.readObject();
        in.close();

        // Test that both validators behave the same way
        String[] testValues = { "valid", null, "", " ", "\t", "\n", "test123" };

        for (String value : testValues) {
            assertEquals(originalValidator.eval(value), deserializedValidator.eval(value),
                    "Validators should behave identically for value: " + value);
        }
    }
}